### ISSUE TITLE
Support python 3.10+

### ISSUE THREAD

**Issue by leonardostefan** [boxed/mutmut#281]
I'm using mutmut in my work projects with python 3.10 and its working except for a single project that we are using the "new" syntax of match-case (PEP 634). In this case, mutmut broke and return 100% of mutants killeds

There's any planning to add full suport for python 3.10?

**Comment by boxed** [comment]
Blocked by https://github.com/davidhalter/parso/issues/138

**Comment by gulliver-madrid** [comment]
While waiting for the full support of Python 3.10, is there any recommended workaround that could be used for projects employing the match-case syntax? Any tips would be greatly appreciated.

**Comment by boxed** [comment]
Sorry, I don't really have anything to say to that. Extracting code you need to be mutation tested and testing it in a separate project is one way at least.. not great but 🤷

**Comment by nkhitrov** [comment]
Hello there! Any news?

**Comment by boxed** [comment]
Read the linked issues for parso

**Comment by nkhitrov** [comment]
Last comment was 2 years ago. It looks like nobody cares about this problem. Maybe change the parser? For example, to [libcst](https://libcst.readthedocs.io/en/latest/index.html)?

**Comment by boxed** [comment]
Yea, that's one way to go for sure. I switched AST library for mutmut once before, and that wasn't as hard as one might expect. I'm very open to PRs, but I don't have much time for this myself.

**Comment by boxed** [comment]
I've also thought about using whatever Ruff uses, or whatever black uses...

**Comment by yonkeltron** [comment]
FYI, there seems to be another issue which I've uncovered coming from parso. I'm not sure if it's the same or not, so please let me know if I should open up a new issue, instead.

This seems to be a parso problem with `import asyncio`:

```console
$ mutmut run

- Mutation testing starting -

These are the steps:
1. A full test suite run will be made to make sure we
   can run the tests successfully and we know how long
   it takes (to detect infinite loops for example)
2. Mutants will be generated and checked

Results are stored in .mutmut-cache.
Print found mutants with `mutmut results`.

Legend for output:
🎉 Killed mutants.   The goal is for everything to end up in this bucket.
⏰ Timeout.          Test suite took 10 times as long as the baseline so were killed.
🤔 Suspicious.       Tests took a long time, but not long enough to be fatal.
🙁 Survived.         This means your tests need to be expanded.
🔇 Skipped.          Skipped.

mutmut cache is out of date, clearing it...
1. Running tests without mutations
⠸ Running...Done
Failed to parse app/cli/commands/relay_events/relay_handlers.py. Internal error from parso follows.
----------------------------------
Traceback (most recent call last):
  File "/virtualenvs/service-f1pOEzwg-py3.12/lib/python3.12/site-packages/parso/parser.py", line 180, in _add_token
    plan = stack[-1].dfa.transitions[transition]
           ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
KeyError: <PythonTokenTypes.NAME: TokenType(NAME)>

During handling of the above exception, another exception occurred:
```

SNIP

```
RuntimeError: Failed while creating mutations for app/cli/commands/relay_events/relay_handlers.py, for line "import asyncio"
```

Is this something new which needs a new issue or is this related to the same parso issue above?

**Comment by FrozenDude101** [comment]
I think this is related to parso, any python file using the "type" keyword added in 3.12 errors:

`parso.parser.ParserSyntaxError: ('SyntaxError: invalid syntax', <ErrorLeaf: PythonTokenTypes.NAME:'Foo', (17, 5)>)`
`The above exception was the direct cause of the following exception:`
`RuntimeError: Failed while creating mutations for pyastodon/models\application.py, for line "from typing import Optional"`

A minimal file is just:
`type Foo = int`

**Comment by jhovell** [comment]
> ... Extracting code you need to be mutation tested and testing it in a separate project is one way at least.. not great but 🤷

I'm surprised the workaround would have to be so painful / impractical. In particular I'm puzzled why excluding files with `match` or other newer keywords doesn't seem to work either excluding offending files via `def pre_mutation(context):` or via `--paths-to-exclude` ... based on adding some debug statements it seems like the error is hit before any files are excluded or `pre_mutation` is run. Would it be hard to make sure files are excluded before they are attempted to be parsed?

**Comment by boxed** [comment]
As it stands right now I am in the process of writing version 3.0 of mutmut which is a big rewrite basically. After this is released, the AST library can be replaced (again). 

I just have to guess and hope that any AST lib I choose to use will be kept up to date. This is the second time I've been burned by this specific problem.

**Comment by boxed** [comment]
Mutmut 3 is released. I did not switch AST library, so maybe that's next on the agenda.

**Comment by jhovell** [comment]
nice! congrats. so just to be clear though that means if there are python 3.10+ keywords like `match` it’ll crash or is there any improvements in workarounds?

**Comment by boxed** [comment]
Yea. It'll still crash on parsing that entire file. I guess it should be pretty simple to make mutmut just skip all files it can't parse.

**Comment by jhovell** [comment]
i would +1 that idea. would improve a major impediment to practical usage in an existing project

**Comment by boxed** [comment]
3.0.2 released with this fix (and another tiny thing that was a showstopper too)
